### PR TITLE
Include interface ID in the PeerAddress that gets handed up from UDP/TCP code

### PR DESCRIPTION
**Pull request by bzbarsky-apple**
#### Problem
When our UDP and TCP code hands a PeerAddress up the stack for incoming messages, it does not include the interface id in that PeerAddress.  When upper layers then update their cached PeerAddress based on the incoming address, they can lose the ability to send messages properly (e.g if sending to an IPv6 link-local address).

#### Change overview
Hand up the right interface ID if we can.  For TCP this is half-baked at best and definitely does not work for LwIP.

#### Testing
Manually tried running `/out/debug/standalone/chip-tool onoff toggle 1` when paired to a device that has an IPv6 link-local address.  Without these changes the message gets sent but the standalone ack for the response gets "No route to host".  With these changes, the ack is also sent correctly.

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]

**Review by Damian-Nordic — APPROVED** [review]

**Comment by woody-apple** [comment]
@msandstedt @andy31415 @jepenven-silabs ? Would love some eyes on this to unblock testing. Thanks!

**Comment by woody-apple** [comment]
@saurabhst ?

**Review by saurabhst — APPROVED** [review]

File: src/inet/TCPEndPoint.cpp [hunk 1]
```diff
@@ -696,6 +696,67 @@ INET_ERROR TCPEndPoint::GetLocalInfo(IPAddress * retAddr, uint16_t * retPort)
...
+    if (sa.any.sa_family == AF_INET6)
+    {
+        if (IPAddress::FromIPv6(sa.in6.sin6_addr).IsIPv6LinkLocal())
+        {
+            *retInterface = sa.in6.sin6_scope_id;
+        }
+        else
+        {
+            // TODO: Is there still a meaningful interface id in this case?
+            *retInterface = INET_NULL_INTERFACEID;
```
Comment (msandstedt) [inline src/inet/TCPEndPoint.cpp:739]: As far as I know, sin6_scope_id will only be an interface ID for link local addresses.  So in short, no, I don't believe an interface ID can be extracted for addresses other than link-local.  And I can't think of a reason why we would need this for non-link-local addresses.  In theory we could bind to a specific interface at which we know a device may reach us.  But I don't know why we would want to or need to.

As to the other possible overloaded meanings of sin6_scope_id (zones or site indices?), it's hard to imagine these would be useful to us, or could even be dealt with in a portable way.  I don't know much about those though, except that they are not interface indices.

TLDR: It looks to me like the code is doing the correct thing here.

Comment (bzbarsky-apple) [reply src/inet/TCPEndPoint.cpp:739]: Thank you for looking into this!  This part I was really not sure about, as you can tell from the comments.  ;)
